[Better Transformer] make is_causal a hint and force attn_mask to be set on `is_causal=True` in F.MHA

mikekgfb opened the pull request:
Summary:
This fixes an issue raised in [is_causal parameter in torch.nn.TransformerEncoderLayer.forward does not work #96941](https://github.com/pytorch/pytorch/issues/96941) where results computed with is_causal do not properly reflect causal masking.

In PyTorch 2.0, Accelerated PT Transformers added the is_causal parameter to legacy nn.Transformer* and nn.MHA APIs aligned with and intended to engage the is_causal parameter of the new scaled_dot_product_attention (SDPA) operator.

At present is_causal works differently for Transformer* modules, compared to the nn.MHA and F.MHA:
* The nn.Transformer* modules treat is_causal as an optional indicator about the format of attn_mask. This is because some layers (such as the CLIP layer use the attention mask in the layer, and thus the attn_mask was a required feature.)
* Initially, nn.MHA and F.MHA were defined to align with F.SDPA in behavior: a user may specify either the attention mask, or is_causal, but not both.  It seemed to make sense at the time to align SDPA and MHA, esp since there was a larger overlap of parameters which have since changed, e.g., with the removal of need_weights from SDPA. (See below for why this makes sense.)

Unfortunately, this does not work because of how MHA was changed to handle the need_weights parameter.  When need_weights is present, we do not (any more) call SDPA because support for need_weights was removed from SDPA before the release.  The rationale is that need_weights defeats all optimization at the foundation of SDPA performance.  Having the flag might thus mislead users into thinking they get good performance and have them disappointed when they enable a legacy feature of MHA which massively degrades performance.  (They might not think anything of enabling that, because it is on by default in MHA today, which leads to more  issues.)

Since SDPA does not (no longer) support need_weights, we need to pick a separate path which implements attention using a set of discrete operations that allocates a tensor for weights.  Alas, this code path does not have support for is_causal, because attention is implemented as matmul and using the attention mask.  Thus, is_causal has no impact.  (A substantially similar situation arises with how kpm is implemented today because Nested Tensors are not supported by torch.compile() in 2.0)

This problem was masked because all uses of legacy nn.MHA (and F.MHA) come through nn.Transformer* which called self-attention (i.e., nn.MHA) only ever with the attention mask attn_mask, and never with is_causal, a missed optimization opportunit that would have been addressed in a future performance update.  

Regrettably, always calling nn.MHA with attn_mask prevented diagnosing of the issue of not having a suitable attention mask when need_weights support was dropped from SDPA and a discrete implementation of attention was added for that scenario, and for the execution path with key_padding_mask. 

We have two options to address this issue:

Solution 1: Whenever nn.MHA and F.MHA are executed with is_causal set, we internally create a causal mask at significant expense of allocating a tensor and filling it with a triangular causal matrix.  This increases memory usage, and runtime, for allocating a causal mask.  To add insult to injury, in all current (and likely future) execution scenarios, MHA is called by a model using the nn.Transformer API which already has that matrix and passes it from nn.module to nn.module.  Then the passing in of attn_mask has to be suppressed by nn.TransformerEncoderLayer, only for nn.MHA to immediately allocate the very same tensor again to satisfy the requirement to have an attention mask for the computation. (We expect new use cases to use SDPA directly.)

Solution 2: We align the behavior of nn.MHA and F.MHA with the rest of the existing nn.Transformer API, and require the attention mask to be passed into nn.MHA in addition to is_causal as an optional indicator about the nature of the attention mask rather than as an alternative to attn_mask.  Then, when we choose the code path for processing MHA with need_weights or a key_padding_mask, we have the attn_mask passed down through the nn.Transformer* hierarchy, without the added overhead of allocating an attention mask as in scenario 1.  

This PR implements solution 2 which offers better performance and in retrospect aligns MHA better with the rest of the Transformer modules as the definition of SDPA evolved into a more streamlined high-performance operator.  It ostensibly changes how is_causal works, by requiring the attention mask to be specified.  However, as described here, and as shown in the submitted issue, is_causal is not working as intended today, so it requires a change regardless.  

In that sense, a change in API does not occur per-se, as the current implementation is not working, and a change has to occur either way to resolve the submitted issue, breaking any use cases that depend on the current implementation.  Checks exist (and more can be added) that flag any scenarios where is_causal is passed as True, but no attention mask is provided, ensuring that there's not quiet change from even the faulty behavior present in 2.0.  

As  an upside, the present implementation will improve performance by addressing the passing of the is_causal flag from Transformer modules to MHA, speeding up training for these examples, e.g., finetuning BERT, RoBERTa, XLM-R models.

Differential Revision: D44245725

albanD commented:
I'm a bit confused by this change.
- Could you add a more detailed description on what is the motivation to do this?
- Isn't this a BC-breaking change? We should follow our [bc policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes).
- Also there are quite a few unrelated comments left? We should make sure to keep the code clean and follow [high level design](https://pytorch.org/docs/stable/community/design.html).

mikekgfb commented:
> I'm a bit confused by this change.
> 
> * Could you add a more detailed description on what is the motivation to do this?
> * Isn't this a BC-breaking change? We should follow our [bc policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes).

This is still exploratory / experimental, and one possible way to resolve Issue #96941. A change here is not backward breaking because the combination of parameters addressed here currently doesn't work in the API that is being changed here, as per the issue.

albanD commented:
> This is still exploratory / experimental

Aren't `nn.Transformer` and associate stable features?

mikekgfb commented:
> > This is still exploratory / experimental
> 
> Aren't `nn.Transformer` and associate stable features?

The reference was to this PR, not to the API.

albanD commented on torch/nn/functional.py:
```diff
@@ -5182,6 +5196,29 @@ def multi_head_attention_forward(
+            #     S = key.size(-2)
+            #     attn_mask = torch.triu(torch.full((L, S), float('-inf'), device=query.device, dtype=query.dtype), diagonal=1)
+
+            assert not (is_causal and attn_mask is None), "FIXME: instantiate causal attn_mask"
```
If this is an error that the user will see, we should have a nicer error message.

mikekgfb replied:
The user should not.  This was a debug check.  We have two options - we can either require attn_mask to always be supplied, or instantiate it in cases where we need.  

I settled on conceptually using is_)causal as a flag about the nature of attn_mask, and always providing attn_mask is conceptually simpler.  (Basically, from a user perspective, you can specify is_causal to get better perf, or just keep using the legacy API.)

albanD commented on torch/nn/functional.py:
```diff
@@ -5301,6 +5338,14 @@ def multi_head_attention_forward(
         B, Nt, E = q.shape
         q_scaled = q / math.sqrt(E)
+
+        # if is_causal and attn_mask is None:
```
debug code?

mikekgfb replied:
yes.

albanD commented on test/test_transformers.py:
```diff
@@ -818,89 +832,6 @@ def test_bias_is_none(self):
         model(x, x, x)
         # completes without error
 
-    @parametrize("device", device_list)
```
Why remove these tests?
Things that used to be errors/unexpected need to be updated but the bulk of the test still makes sense no? Or this is already done somewhere else?

mikekgfb replied:
the removed checks were doing argument checks on combinations of inputs.  Removed them initially to not repeatedly have to change them as I explore solutions.  (github was the quickest way to run all tests, so I submitted PR revisions on every rev I wanted to test....)

In retrospect I wonder whether it's worthwhile to run all the parameter combination tests.

albanD replied:
Testing all combination of parameters is definitely the way to go to make sure we don't regress on corner cases.
In the grand scheme of things, we should have ModuleInfo testing for this that will allow us to do this easily but I think this is fine for now.

Adding back the updated tests here sounds good.

mikekgfb commented on torch/nn/functional.py:
```diff
@@ -5066,10 +5066,17 @@ def multi_head_attention_forward(
             the batches while a 3D mask allows to specify a different mask for the entries of each batch.
         is_causal: If specified, applies a causal mask as attention mask, and ignores
             attn_mask for computing scaled dot product attention.
+            *Warning*: If is_causal is set to True and a non-causal attention
```
Good way to frame this.  The "Warning" was when you claim you have a causal mask in attn_mask using is_causal, but it is not.

mikekgfb commented on torch/nn/functional.py:
```diff
@@ -5170,8 +5177,15 @@ def multi_head_attention_forward(
-    if is_causal:
-        attn_mask = None
+    if is_causal and attn_mask is None:
+        raise RuntimeError("Need attn mask if specifying is_causal")
```
Good point!m We actually have a static method as a member of transformer class - generate_square_subsequent_mask L 152 transformer.py

mikekgfb commented on torch/nn/modules/activation.py:
```diff
@@ -1097,8 +1097,6 @@ def forward(
         .. note::
             `batch_first` argument is ignored for unbatched inputs.
         """
-        if attn_mask is not None and is_causal:
```
good catch.

albanD reviewed (approved):
SGTM Just needs a rebase!

mikekgfb commented:
@pytorchbot merge

Commit:
is causal transformer - force attn_mask to be set on is_causal in F.MHA (#97214)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/97214

is causal transformer - force attn_mask to be set on is_causal in F.MHA
(and transitively nn.MHA)

Make attn_mask compulsory on is_causal

Test Plan: sandcastle / github

Differential Revision: D44245725

fbshipit-source-id: 9d7640fc2fc2c11fc90b9c47ac4a0df8eed57f63

mxkrn commented:
I understand the clarification, but it would be great if the documentation was updated accordingly - currently it still says that option 1 is the current behavior.

Docs from latest `main` (https://pytorch.org/docs/main/generated/torch.nn.TransformerEncoder.html?highlight=transformerencoder#torch.nn.TransformerEncoder) say:

> is_causal ([Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[bool](https://docs.python.org/3/library/functions.html#bool)]) – If specified, applies a causal mask as mask (optional) and ignores attn_mask for computing scaled dot product attention. Default: False.

mikekgfb commented:
Docs are extracted from doc strings when a new version is released. Please check the doc strings in the diff whether you consider them clear in this respect.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Max Kraan ***@***.***>
Sent: Thursday, April 27, 2023 6:45:47 AM
To: pytorch/pytorch ***@***.***>
Cc: Michael Gschwind ***@***.***>; Author ***@***.***>
Subject: Re: [pytorch/pytorch] [Better Transformer] make is_causal a hint and force attn_mask to be set on `is_causal=True` in F.MHA (PR #97214)

I understand the clarification, but it would be great if the documentation was updated accordingly - currently it still says that option 1 is the current behavior. Docs from latest main (https: //pytorch. org/docs/main/generated/torch. nn. TransformerEncoder. html?highlight=transformerencoder#torch. nn. TransformerEncoder)
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender

ZjQcmQRYFpfptBannerEnd

I understand the clarification, but it would be great if the documentation was updated accordingly - currently it still says that option 1 is the current behavior.

Docs from latest main (https://pytorch.org/docs/main/generated/torch.nn.TransformerEncoder.html?highlight=transformerencoder#torch.nn.TransformerEncoder<https://pytorch.org/docs/main/generated/torch.nn.TransformerEncoder.html?highlight=transformerencoder#torch.nn.TransformerEncoder>) say:

is_causal (Optional<https://docs.python.org/3/library/typing.html#typing.Optional>[bool<https://docs.python.org/3/library/functions.html#bool>]) – If specified, applies a causal mask as mask (optional) and ignores attn_mask for computing scaled dot product attention. Default: False.

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/97214#issuecomment-1525724173>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOT4XHPVVUA3OXQM6L5OHS3XDJ2AXANCNFSM6AAAAAAWB55TOQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>
